### ISSUE TITLE
Advancing travis testing

### ISSUE THREAD

**Issue by simonclausen** [dnscryptio/dnscrypt-wrapper-docker#1]
The [dnscrypt-proxy-docker](https://github.com/dnscryptio/dnscrypt-proxy-docker) image is now being build and put on [dnscryptio/dnscrypt-proxy-docker](https://hub.docker.com/r/dnscryptio/dnscrypt-proxy-docker/).

This means we now have an additional building block to do some more advanced testing of dnscrypt-wrapper-docker.

But how do we manage keys in the wrapper image for test?

As I see it right now, we have two options:
- Build and hardcode some keys into the image to be used as default, if no volume is mounted 
- Use the a wrapper container in the build to generate the keys and setup proxy with them before running final tests

First option is unpretty and could result in a poor security if used by someone who does not know that keys need to be supplied externally. I would prefer not to do this, but then again it looks like a shortcut way to getting a more advanced test working.

Option two looks good on the surface. But how does one change directory to /opt/dnscrypt-wrapper/etc/keys/ for the key and cert generating commands? Use WORKDIR? Is that pretty?

Quick overview of what needs to be done:
- Generate provider keypair (and save fingerprint/pubkey to be used with proxy for test)
- Generate time-limited key + cert
- Start the wrapper service
- Start proxy with the info for wrapper

(This issue is mostly a note, since I won't be able to learn and make all of this tonight!)

**Comment by simonclausen** [comment]
Come to think of it, keys could just be provided by a wget from somewhere we place some test keys. No need to modify the image with default keys or generate them during the build test.

I would like a suggestion on how the last option could be done, though.

**Comment by kidmose** [comment]
I'm for generating keys while building. 
Testing should have minimal external dependency, eg. web server with keys, otherwise we'll be pulling our hair because of sporadic failure some day. Tests need to be deterministic whenever possible. 

Can we do some magic by mounting volumes (docker -v) ? Maybe get symlink-like effects by mounting the same host folder two places in the docker image ? (And between wrapper and docker) That enables us to generate keys and also use them in another images.

**Comment by kidmose** [comment]
Likely crazy idea: could test keys be part of the Travis environment variables? (We still need to test key generating capabilities, but one functionality pr. test case I highly desirable, when planning for not pulling our your own hair.

**Comment by simonclausen** [comment]
It should actually be possible to set keys as part of the travis build env. Good idea. That may be a quick way to get started.

The commands for generating keys in dnscrypt-wrapper assumes you want the keys placed in the current working directory. So I guess it should be done like this:

`docker run -w "/opt/dnscrypt-wrapper/etc/keys" -v $KEYDIR:/opt/dnscrypt-wrapper/etc/keys dnscrypt-wrapper --gen-provider-keypair` (etc...)

How do we best get and save the public key fingerprint? It looks like `--show-provider-publickey-fingerprint` does not actually work:

```
simons-mbp:dnscrypt-wrapper-docker simon$ docker run -w "/opt/dnscrypt-wrapper/etc/keys" dnscrypt-wrapper --show-provider-publickey-fingerprint --provider-publickey-file="/opt/dnscrypt-wrapper/etc/keys/public.key"
error: unknown option `--show-provider-publickey-fingerprint`
Usage: dnscrypt-wrapper [options]
```

Interestingly, it seems like there are already keys present in the freshly built image:

```
simons-mbp:dnscrypt-wrapper-docker simon$ docker run -w "/opt/dnscrypt-wrapper/etc/keys" dnscrypt-wrapper --gen-provider-keypair
Generate provider key pair... ok.
Public key fingerprint: A90E:D9ED:24B4:5D02:4FF5:B178:AF86:4D31:2FD9:F8F0:3FFF:65F7:C2DE:1252:8A13:8FD0

This is the provider key you should give to users for your service.
(i.e. dnscrypt-proxy --provider-key=A90E:D9ED:24B4:5D02:4FF5:B178:AF86:4D31:2FD9:F8F0:3FFF:65F7:C2DE:1252:8A13:8FD0
                     --resolver-address=<your resolver public IP>
                     --provider-name=2.dnscrypt-cert...)


*KEYS HAVE NOT BEEN SAVED*

A provider key pair already exists.
If you really want to overwrite it, delete the public.key and secret.key files.

The provider public key is what client give to dnscrypt-proxy
in order to use your service (long-term key).

Unless the private key has been compromised, you probably do not
want to overwrite it with a new one.

Usually, what you want (if current certificates are about to expire)
to regenerate is the server key pairs (--gen-crypt-keypair),
not the master keys.
```

**Comment by simonclausen** [comment]
~~Nm on the last point. It looks like it's just me not thinking about the key dir only being writable when a volume is mounted and dnscrypt-wrapper defaulting to a somewhat misleading error message.~~

~~Fun times learning this docker thing :p~~

Nope, seems like something else is up. When mounting a volume explicitly, same result happens. Same when just letting docker take care of volume location on host (-v /opt/dnscrypt-wr...).

**Comment by simonclausen** [comment]
```
simons-mbp:dnscrypt-wrapper-docker simon$ echo $KEYDIR
/tmp/dnscrypttest
simons-mbp:dnscrypt-wrapper-docker simon$ docker run -w "/opt/dnscrypt-wrapper/etc/keys" -v $KEYDIR:/opt/dnscrypt-wrapper/etc/keys dnscrypt-wrapper --gen-provider-keypair
Generate provider key pair... ok.
Public key fingerprint: EB8D:F733:13B2:BA02:225B:8670:61C7:0159:B78C:3F51:1557:AC5F:E484:6729:86CB:6C25

This is the provider key you should give to users for your service.
(i.e. dnscrypt-proxy --provider-key=EB8D:F733:13B2:BA02:225B:8670:61C7:0159:B78C:3F51:1557:AC5F:E484:6729:86CB:6C25
                     --resolver-address=<your resolver public IP>
                     --provider-name=2.dnscrypt-cert...)


*KEYS HAVE NOT BEEN SAVED*

A provider key pair already exists.
If you really want to overwrite it, delete the public.key and secret.key files.

The provider public key is what client give to dnscrypt-proxy
in order to use your service (long-term key).

Unless the private key has been compromised, you probably do not
want to overwrite it with a new one.

Usually, what you want (if current certificates are about to expire)
to regenerate is the server key pairs (--gen-crypt-keypair),
not the master keys.
simons-mbp:dnscrypt-wrapper-docker simon$ ls /tmp/dnscrypttest
```

```
simons-mbp:dnscrypt-wrapper-docker simon$ docker run -w "/opt/dnscrypt-wrapper/etc/keys" -v /opt/dnscrypt-wrapper/etc/keys dnscrypt-wrapper --gen-provider-keypair
Generate provider key pair... ok.
Public key fingerprint: 0297:C637:E509:2C54:8331:D39F:13FD:E19A:9DA0:F67E:8741:56B9:677D:7AB3:EA61:A3D8

This is the provider key you should give to users for your service.
(i.e. dnscrypt-proxy --provider-key=0297:C637:E509:2C54:8331:D39F:13FD:E19A:9DA0:F67E:8741:56B9:677D:7AB3:EA61:A3D8
                     --resolver-address=<your resolver public IP>
                     --provider-name=2.dnscrypt-cert...)


*KEYS HAVE NOT BEEN SAVED*

A provider key pair already exists.
If you really want to overwrite it, delete the public.key and secret.key files.

The provider public key is what client give to dnscrypt-proxy
in order to use your service (long-term key).

Unless the private key has been compromised, you probably do not
want to overwrite it with a new one.

Usually, what you want (if current certificates are about to expire)
to regenerate is the server key pairs (--gen-crypt-keypair),
not the master keys.
```

Is it docker(-tools) on mac playing with me?

**Comment by kidmose** [comment]
Something is certainly amis. Which container are you running? Did we commit keys from travis? Maybe discard local image and pull from docker hub.

**Comment by simonclausen** [comment]
We don't have dnscrypt-wrapper-proxy on docker hub. I would like the testing to be a bit better before we start pushing there.

Anyway, I am building the Dockerfile in this repo and running it. I am not sure what you mean by keys from travis..

**Comment by kidmose** [comment]
I had gotten the impression that you had built and pushed to docker hub. 
I figured you might have committed an image containing the keys resulting from doing keygen. 
In that case the above result would be expected.
Anyway, I was off :P

**Comment by simonclausen** [comment]
Well, the immediate issue is that key generation  is not working (see errors above). 

Has it ever worked for you?

**Comment by kidmose** [comment]
Work locally for me:

```
/home/.../dnscrypt-wrapper-docker $ KEYDIR=/tmp/dnscrypt-wrapper-keys
/home/.../dnscrypt-wrapper-docker $ rm -rf $KEYDIR; mkdir $KEYDIR
/home/.../dnscrypt-wrapper-docker $ docker run --interactive --tty --rm --volume $KEYDIR:/etc/dnscrypt-wrapper/cert -w /etc/dnscrypt-wrapper/cert dnscrypt-wrapper-docker --gen-provider-keypair
Generate provider key pair... ok.
Public key fingerprint: 88A8:DB41:FE68:D0DD:5F6E:39F2:AF18:91F5:8370:C8D2:F659:284F:FB9F:8DE9:4361:1197

This is the provider key you should give to users for your service.
(i.e. dnscrypt-proxy --provider-key=88A8:DB41:FE68:D0DD:5F6E:39F2:AF18:91F5:8370:C8D2:F659:284F:FB9F:8DE9:4361:1197
                     --resolver-address=<your resolver public IP>
                     --provider-name=2.dnscrypt-cert...)
Keys are stored in public.key & secret.key.
/home/.../dnscrypt-wrapper-docker $ ls $KEYDIR 
public.key  secret.key
```

**Comment by simonclausen** [comment]
In this I've generated the relevant dnscrypt keys and certs via a dnscrypt-wrapper install on the host (my macbook):

```
simons-mbp:dnscrypttest simon$ pwd
/Users/simon/dnscrypttest
simons-mbp:dnscrypttest simon$ ls -la
total 32
drwxr-xr-x   6 simon  staff   204 Sep 11 11:11 .
drwxr-xr-x+ 52 simon  staff  1768 Sep 11 11:09 ..
-r--r--r--   1 simon  staff    32 Sep 11 11:11 crypt_secret.key
-r--r--r--   1 simon  staff   124 Sep 11 11:11 dnscrypt.cert
-r--r--r--   1 simon  staff    32 Sep 11 11:07 public.key
-r--r--r--   1 simon  staff    64 Sep 11 11:07 secret.key
simons-mbp:dnscrypt-wrapper-docker simon$ KEYDIR=/Users/simon/dnscrypttest/
simons-mbp:dnscrypt-wrapper-docker simon$ docker run --volume=$KEYDIR:/opt/dnscrypt-wrapper/etc/keys --publish=127.0.0.1:4443:4443/udp dnscrypt-wrapper --listen-address=0.0.0.0:4443 --resolver-address=8.8.8.8:53 --provider-name=2.dnscrypt-cert.example.com --provider-cert-file=/opt/dnscrypt-wrapper/etc/keys/dnscrypt.cert --crypt-secretkey-file=/opt/dnscrypt-wrapper/etc/keys/crypt_secret.key
```

dnscrypt-wrapper in the docker container starts and runs without issue.

> I've come across this info in the documentation:
> If you are using Docker Machine on Mac or Windows, your Docker Engine daemon has only limited access to your OS X or Windows filesystem.
> (https://docs.docker.com/engine/tutorials/dockervolumes/)

No words about Docker Toolbox, but I'm guessing it's a similar issue. In the previous tests I used /tmp/dnscrypttest on the host. This may be the issue.

I will attempt to create keys using /Users/simon/dnscrypttest/ as the mounted volume.

Edit: I pasted the wrong terminal output, corrected now

**Comment by simonclausen** [comment]
```
simons-mbp:dnscrypt-wrapper-docker simon$ docker run --volume=$KEYDIR:/opt/dnscrypt-wrapper/etc/keys dnscrypt-wrapper --gen-provider-keypair
Generate provider key pair... ok.
Public key fingerprint: 1C74:ACD3:3015:8CD5:8F15:CE5F:5062:B71D:EDCB:3443:C8DC:9030:7932:D141:5CBA:EBB1

This is the provider key you should give to users for your service.
(i.e. dnscrypt-proxy --provider-key=1C74:ACD3:3015:8CD5:8F15:CE5F:5062:B71D:EDCB:3443:C8DC:9030:7932:D141:5CBA:EBB1
                     --resolver-address=<your resolver public IP>
                     --provider-name=2.dnscrypt-cert...)


*KEYS HAVE NOT BEEN SAVED*

A provider key pair already exists.
If you really want to overwrite it, delete the public.key and secret.key files.

The provider public key is what client give to dnscrypt-proxy
in order to use your service (long-term key).

Unless the private key has been compromised, you probably do not
want to overwrite it with a new one.

Usually, what you want (if current certificates are about to expire)
to regenerate is the server key pairs (--gen-crypt-keypair),
not the master keys.
simons-mbp:dnscrypt-wrapper-docker simon$ echo $KEYDIR
/Users/simon/dnscrypttest/
simons-mbp:dnscrypt-wrapper-docker simon$ ls -la /Users/simon/dnscrypttest/
total 0
drwxr-xr-x   2 simon  staff    68 Sep 11 11:36 .
drwxr-xr-x+ 52 simon  staff  1768 Sep 11 11:09 ..
```

I think this can be put down to how toolbox work - or the opposite - on OS X.

**Comment by kidmose** [comment]
I've got key gen working in this branch, including extraction of fingerprint: https://github.com/kidmose/dnscrypt-wrapper-docker/blob/feature/travis-run-wrapper/.travis.yml  

It also starts the wrapper, and tries to run dnscrypt-proxy for testing, but this fails, likely because I mess something up with the parameters for proxy:
https://travis-ci.org/kidmose/dnscrypt-wrapper-docker/builds/159125758

**Comment by simonclausen** [comment]
Looks to be because dnscrypt-proxy rejects "2.dnscrypt-cert.example.com".

This works (dnscrypt.eu-dk):

```
$ docker run -e RESOLVER_ADDR=77.66.84.233:443 -e PROVIDER_NAME=2.dnscrypt-cert.resolver2.dnscrypt.eu -e PROVIDER_KEY=3748:5585:E3B9:D088:FD25:AD36:B037:01F5:520C:D648:9E9A:DD52:1457:4955:9F0A:9955 dnscryptio/dnscrypt-proxy-docker
[NOTICE] Starting dnscrypt-proxy 1.7.0
[INFO] Generating a new session key pair
[INFO] Done
[INFO] Server certificate with serial '0001' received
[INFO] This certificate is valid
[INFO] Chosen certificate #808464433 is valid from [2016-09-10] to [2017-09-10]
[INFO] Server key fingerprint is E65D:5D0F:1846:A060:402C:46FD:DDE3:1789:F1CB:143C:EB34:7008:372F:E9A8:480D:1E46
[NOTICE] Proxying from 0.0.0.0:53 to 77.66.84.233:443
```

This does not (dnscrypt.eu-dk w/ modified dnscrypt-cert record):

```
$ docker run -e RESOLVER_ADDR=77.66.84.233:443 -e PROVIDER_NAME=2.dnscrypt-cert.example.com -e PROVIDER_KEY=3748:5585:E3B9:D088:FD25:AD36:B037:01F5:520C:D648:9E9A:DD52:1457:4955:9F0A:9955 dnscryptio/dnscrypt-proxy-docker
[NOTICE] Starting dnscrypt-proxy 1.7.0
[INFO] Generating a new session key pair
[INFO] Done
[ERROR] Unable to retrieve server certificates
```

But then again wrapper docs says the following:

> The provider name can be anything; it doesn't have to be within an existing domain name. However, it has to start with 2.dnscrypt-cert..
> 
> When the service is started with the --provider-cert-file switch, the proxy will automatically serve the certificate as a TXT record when a query for the provider name is received.

**Comment by simonclausen** [comment]
I've been doing some more local testing: as long as the provider name is set to the same thing on the wrapper and proxy, a connection is made. As the documentation notes, the wrapper serves the certificate when asked for the TXT record by a client.

How this can take place before the cert is "retrieved", I don't know. But I think that's not immediately relevant for this issue.

Also:
- Everything in the parameters handed over via the travis config looks good and works in a local test setup.
- I've tried different methods of running dnscrypt-proxy in the dnscrypt-proxy-docker image - works locally, goes bad in travis build.
- I've added a setting options for loglevel in dnscrypt-proxy-docker, and set it to max, hoping to get more info from the build, but to no result.

**Comment by simonclausen** [comment]
I just re-read everything in this thread. It's turning out to be pretty embarrassing :)

I got from this:

```
[ERROR] Resolver information required.
[ERROR] The easiest way to do so is to provide a resolver name.
[ERROR] Example: dnscrypt-proxy -R mydnsprovider
[ERROR] See the file [/opt/dnscrypt-proxy/share/dnscrypt-proxy/dnscrypt-resolvers.csv] for a list of compatible public resolvers
[ERROR] The name is the first column in this table.
[ERROR] Alternatively, an IP address, a provider name and a provider key can be supplied.
[ERROR] Please consult https://dnscrypt.org and the dnscrypt-proxy(8) man page for details.
```

To this:

```
[NOTICE] Starting dnscrypt-proxy 1.7.0
[INFO] Ephemeral keys enabled - generating a new seed
[INFO] Done
[ERROR] Unable to retrieve server certificates
```

... I guess that's some kind of progress.

Now, let's look at what's happening. I am telling dnscrypt-proxy to connect to -wrapper on 127.0.0.1:4443. Wouldn't it try to connect to localhost within the container? I think so.

So linking the containers should solve it. I'll go do that.

**Comment by simonclausen** [comment]
As discussed today I just spun up a Ubuntu 14.04 VM on AWS, installed docker, got the current Dockerfile from the repo, and ran each and every step in the current .travis.yml commit (https://github.com/dnscryptio/dnscrypt-wrapper-docker/pull/4/commits/db19a3a5c2aed208940131e512f624f54e31cda1).

Guess what: dnscrypt-wrapper-docker seems to run. But dnscrypt-proxy-docker does not:

```
docker run -e RESOLVER_ADDR=$WRAPPERIP:4443 -e PROVIDER_NAME=2.dnscrypt-cert.test.dnscrypt.io -e PROVIDER_KEY=$PUBLIC_KEYPAIR_FINGERPRINT -e LOGLEVEL=7 -p 127.0.0.1:53:53/udp dnscryptio/dnscrypt-proxy-docker
[ERROR] Resolver information required.
[ERROR] The easiest way to do so is to provide a resolver name.
[ERROR] Example: dnscrypt-proxy -R mydnsprovider
[ERROR] See the file [/opt/dnscrypt-proxy/share/dnscrypt-proxy/dnscrypt-resolvers.csv] for a list of compatible public resolvers
[ERROR] The name is the first column in this table.
[ERROR] Alternatively, an IP address, a provider name and a provider key can be supplied.
[ERROR] Please consult https://dnscrypt.org and the dnscrypt-proxy(8) man page for details.
```

So, more or less the opposite of what's happening on travis-ci!

**Comment by simonclausen** [comment]
N/m on the last point - setting fingerprint had gone bad (didn't scrub single-quotes):

```
# PUBLIC_KEYPAIR_FINGERPRINT=D4D7:31E0:C230:13B6:009F:8AA7:BCE8:E593:223C:2E61:946C:F786:AA40:F8D1:F26E:6245
# docker run -e RESOLVER_ADDR=$WRAPPERIP:4443 -e PROVIDER_NAME=2.dnscrypt-cert.test.dnscrypt.io -e PROVIDER_KEY=$PUBLIC_KEYPAIR_FINGERPRINT -e LOGLEVEL=7 -p 127.0.0.1:53:53/udp dnscryptio/dnscrypt-proxy-docker
[NOTICE] Starting dnscrypt-proxy 1.7.0
[INFO] Ephemeral keys enabled - generating a new seed
[INFO] Done
[INFO] Server certificate with serial #1474309234 received
[INFO] This certificate is valid
[INFO] Chosen certificate #1474309234 is valid from [2016-09-19] to [2017-09-19]
[INFO] Server key fingerprint is 80C6:DF88:DCC3:4FAE:A4B7:A259:21EB:7FC1:4213:9456:DB8F:71F1:F9EE:3FE1:2E34:D01F
[NOTICE] Proxying from 0.0.0.0:53 to 172.17.0.2:4443
```

**Comment by simonclausen** [comment]
Oh, and:

```
$ dig @127.0.0.1 google.com

; <<>> DiG 9.9.5-3ubuntu0.8-Ubuntu <<>> @127.0.0.1 google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27493
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.com.            IN  A

;; ANSWER SECTION:
google.com.     299 IN  A   172.217.16.206

;; Query time: 12 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Mon Sep 19 18:47:19 UTC 2016
;; MSG SIZE  rcvd: 55
```

**Comment by simonclausen** [comment]
There we go: https://github.com/dnscryptio/dnscrypt-wrapper-docker/pull/4/commits/ab4474f2bff7bb92fd321847d9028940ff8f8bb0 / https://travis-ci.org/dnscryptio/dnscrypt-wrapper-docker/builds/161153430

It seems it was the line breaks in the final dnscrypt-wrapper cmd. Go figure.

**Comment by simonclausen** [comment]
That was fun! Closed by #4 - thanks for the help!

**Comment by kidmose** [comment]
Nice work 😊
